### PR TITLE
feat: create SafehillOTP.kt to generate the otp

### PR DESCRIPTION
**Pull request by DikenMaharjan**

### CODE REVIEW

File: src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+class SafehillOTP(
+    digits: Int,
+    validDuration: Duration
+) {
+    private val halfTimeStep = validDuration / 2
+
+    private val otpConfig = TimeBasedOneTimePasswordConfig(
+        codeDigits = digits,
+        hmacAlgorithm = HmacAlgorithm.SHA1,
+        timeStep = halfTimeStep.inWholeMicroseconds,
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt]: Why did you pick microseconds and not milliseconds?

File: src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        )
+    }
+
+
+    /**
+     * @param secret the secret that was used to generate the otp code.
+     * @param code the code to check whether it is valid or not
+     * @return true if the code is valid, false otherwise
+     */
+    fun isValid(
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt]: I wouldn't bake this split logic here, but in the client of this API

File: src/test/kotlin/com/safehill/kcrypto/CipherTests.kt [hunk 3]
```diff
@@ -1,110 +1,133 @@
...
+
+            delay(0.99.seconds)
+
+            assert(SafehillOTP(digits = 6, validDuration = 2.seconds).isValid(newSecret, code11))
+        }
+
+    }
+
+    @Test
+    fun `test otp generated is valid for minimum of half the time step`() {
```
Comment (gennaro-safehill) [inline src/test/kotlin/com/safehill/kcrypto/CipherTests.kt]: Can you make sure the unit tests cover generation and verification across all possible cases?
Think of where generation and verification lay in the timeline and intervals you drew on the board yesterday:
- both generation and verification in first interval
- generation in first and verification in second
- generation in first and verification after second
- etc.

File: src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt (hunk 2 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt]: Do I add a slot parameter to generate the OTP?
O refering to the current time slot.
1 refering to the next slot. 
-1 refering to the previous slot. 

Should the logic we discussed yesterday (where we generate otp valid for x/2 duration and check current slot and previous slot) reside in the client side as well?

File: src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt (hunk 1 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt]: I tried seconds for the first time and generated a code which was valid for 1 seconds.  In this case, I was generating otp which was valid for 1/2 = 0.5 seconds which was being rounded to 0 seconds and the tests were failing.
Then I switched to milliseconds and at the last moment I thought why not use microseconds and increase the precision and thus, it was just a moment of decision to use microseconds.
But upon reviewing the source code of TimeBasedOneTimePasswordConfig, it seems they are working with milliseconds as well. 
Do you want me to refactor it?

File: src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt (hunk 2 above)
Comment (gennaro-safehill) [reply src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt]: Yes, and you can pass the instant

File: src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt (hunk 1 above)
Comment (gennaro-safehill) [reply src/main/kotlin/com/safehill/kcrypto/SafehillOTP.kt]: I would use milliseconds, yes

**Review by gennaro-safehill — COMMENTED** [review]
Replied

File: src/test/kotlin/com/safehill/kcrypto/CipherTests.kt [hunk 4]
```diff
@@ -408,7 +430,13 @@ class CipherTests {
         val signedDigest = Base64.getDecoder().decode(solvedChallenge.signedDigest)
 
         // SERVER verifies the solved challenge
-        /**/ assert(SafehillSignature.verify(challenge, signedChallenge, clientUser.publicSignature))
+        /**/ assert(
```
Comment (gennaro-safehill) [inline src/test/kotlin/com/safehill/kcrypto/CipherTests.kt]: Can remove the `/**/`

File: src/test/kotlin/com/safehill/kcrypto/CipherTests.kt [hunk 5]
```diff
@@ -1,112 +1,133 @@
...
-        )
-        Thread.sleep(max(0, valid11 - 100))
-        val (code12, _) = SafehillCypher.generateOTPCode(
-            secret = newSecret,
-            digits = 6,
-            timeStepInSeconds = 2
-        )
-        assertEquals(code11, code12)
+    @Test
+    fun `test generated code is different after valid time`() {
```
Comment (gennaro-safehill) [inline src/test/kotlin/com/safehill/kcrypto/CipherTests.kt:111]: What's the difference between this and `test generated code is different after valid time has elapsed`?

Comment (DikenMaharjan) [reply src/test/kotlin/com/safehill/kcrypto/CipherTests.kt:111]: Its objectively the same thing. In one case we are delaying the time till we reach the second slot. In another, we are programmatically increasing the Instant to the second slot.

**Comment by DikenMaharjan** [comment]
@gennaro-safehill .. Could you review these changes so that we could merge them?

**Review by gennarinoos — APPROVED** [review]
LGTM

**Pull request merged**